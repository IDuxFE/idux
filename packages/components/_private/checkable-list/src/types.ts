/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

import type { VirtualScrollToFn } from '@idux/cdk/scroll'
import type { ExtractInnerPropTypes, ExtractPublicPropTypes, MaybeArray, VKey } from '@idux/cdk/utils'
import type { DefineComponent, HTMLAttributes, PropType } from 'vue'

export interface CheckableListData {
  key?: VKey
  label?: string
  disabled?: boolean
  [key: string]: unknown
}

export interface CheckableListScroll {
  height?: string | number
  fullHeight?: boolean
}

export const checkableListProps = {
  dataSource: Array as PropType<CheckableListData[]>,
  checkable: {
    type: Boolean,
    default: true,
  },
  removable: {
    type: Boolean,
    default: false,
  },
  checked: Function as PropType<(item: CheckableListData) => boolean>,
  customAdditional: { type: Function as PropType<CheckableListCustomAdditional>, default: undefined },
  disabled: Function as PropType<(item: CheckableListData) => boolean>,
  getKey: Function as PropType<(item: CheckableListData) => VKey>,
  labelKey: String,
  virtual: {
    type: Boolean,
    default: false,
  },
  scroll: Object as PropType<CheckableListScroll>,
  onCheckChange: [Function, Array] as PropType<MaybeArray<(item: CheckableListData, checked: boolean) => void>>,
  onRemove: [Function, Array] as PropType<MaybeArray<(item: CheckableListData) => void>>,
  onScroll: [Function, Array] as PropType<MaybeArray<(evt: Event) => void>>,
  onScrolledChange: [Function, Array] as PropType<
    MaybeArray<(startIndex: number, endIndex: number, visibleData: unknown[]) => void>
  >,
  onScrolledBottom: [Function, Array] as PropType<MaybeArray<() => void>>,
} as const

export const checkableListItemProps = {
  checked: {
    type: Boolean,
    default: false,
  },
  checkable: {
    type: Boolean,
    default: true,
  },
  removable: {
    type: Boolean,
    default: false,
  },
  disabled: {
    type: Boolean,
    default: false,
  },
  label: String,
  value: {
    type: [String, Number, Symbol] as PropType<VKey>,
    required: true,
  },
  onCheckChange: [Function, Array] as PropType<MaybeArray<(checked: boolean) => void>>,
  onRemove: [Function, Array] as PropType<MaybeArray<() => void>>,
} as const

export interface CheckableListApi {
  scrollTo: VirtualScrollToFn
}

export type CheckableListProps = ExtractInnerPropTypes<typeof checkableListProps>
export type CheckableListItemProps = ExtractInnerPropTypes<typeof checkableListItemProps>
export type CheckableListPublicProps = ExtractPublicPropTypes<typeof checkableListProps>
export type CheckableListComponent = DefineComponent<
  Omit<HTMLAttributes, keyof CheckableListPublicProps> & CheckableListPublicProps,
  CheckableListApi
>
export type CheckableListInstance = InstanceType<DefineComponent<CheckableListProps, CheckableListApi>>

export type CheckableListCustomAdditional = (options: {
  data: CheckableListData
  index: number
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
}) => Record<string, any> | undefined
