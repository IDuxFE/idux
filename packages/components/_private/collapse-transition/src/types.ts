/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

import type { ExtractInnerPropTypes, ExtractPublicPropTypes } from '@idux/cdk/utils'
import type { DefineComponent, HTMLAttributes } from 'vue'

import { IxPropTypes } from '@idux/cdk/utils'

export type CollapseTransitionMode = 'height' | 'width'

export const collapseTransitionProps = {
  appear: IxPropTypes.bool.def(false),
  name: IxPropTypes.string,
  mode: IxPropTypes.oneOf<CollapseTransitionMode>(['height', 'width']).def('height'),
  onAfterEnter: IxPropTypes.emit(),
  onAfterLeave: IxPropTypes.emit(),
}

export type CollapseTransitionProps = ExtractInnerPropTypes<typeof collapseTransitionProps>
export type CollapseTransitionPublicProps = ExtractPublicPropTypes<typeof collapseTransitionProps>
export type CollapseTransitionComponent = DefineComponent<
  Omit<HTMLAttributes, keyof CollapseTransitionPublicProps> & CollapseTransitionPublicProps
>
export type CollapseTransitionInstance = InstanceType<DefineComponent<CollapseTransitionProps>>
