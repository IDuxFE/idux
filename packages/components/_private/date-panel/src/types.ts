/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

import type { ExtractInnerPropTypes, ExtractPublicPropTypes, MaybeArray } from '@idux/cdk/utils'
import type { DefineComponent, HTMLAttributes, PropType } from 'vue'

export const datePanelProps = {
  cellTooltip: Function as PropType<(cell: { value: Date; disabled: boolean }) => string | void>,
  disabledDate: Function as PropType<(date: Date) => boolean>,
  type: {
    type: String as PropType<DatePanelType>,
    default: 'date',
  },
  value: [Date, Array] as PropType<Date | (Date | undefined)[]>,
  activeDate: Date,
  visible: {
    type: Boolean,
    default: undefined,
  },
  isSelecting: {
    type: Boolean,
    default: undefined,
  },

  onCellClick: [Function, Array] as PropType<MaybeArray<(date: Date) => void>>,
  onCellMouseenter: [Function, Array] as PropType<MaybeArray<(date: Date) => void>>,
  'onUpdate:activeDate': [Function, Array] as PropType<MaybeArray<(date: Date) => void>>,
}

export type DatePanelProps = ExtractInnerPropTypes<typeof datePanelProps>
export type DatePanelPublicProps = ExtractPublicPropTypes<typeof datePanelProps>
export type DatePanelComponent = DefineComponent<
  Omit<HTMLAttributes, keyof DatePanelPublicProps> & DatePanelPublicProps
>
export type DatePanelInstance = InstanceType<DefineComponent<DatePanelProps>>

export type DatePanelType = 'date' | 'week' | 'month' | 'quarter' | 'year'

// private
export const panelRowProps = {
  rowIndex: {
    type: Number,
    required: true,
  },
}

export const panelCellProps = {
  rowIndex: {
    type: Number,
    required: true,
  },
  cellIndex: {
    type: Number,
    required: true,
  },
}
