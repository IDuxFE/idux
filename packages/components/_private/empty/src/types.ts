/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

import type { EmptyProps as IxEmptyProps } from '@idux/components/empty'

export interface EmptyProps {
  empty?: string | IxEmptyProps
}
