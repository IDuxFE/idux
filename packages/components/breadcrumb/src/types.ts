/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

import type { ExtractInnerPropTypes, ExtractPublicPropTypes } from '@idux/cdk/utils'
import type { DefineComponent, HTMLAttributes } from 'vue'

export const breadcrumbProps = {
  separator: {
    type: String,
    default: '/',
  },
} as const

export type BreadcrumbProps = ExtractInnerPropTypes<typeof breadcrumbProps>
export type BreadcrumbPublicProps = ExtractPublicPropTypes<typeof breadcrumbProps>
export type BreadcrumbComponent = DefineComponent<
  Omit<HTMLAttributes, keyof BreadcrumbPublicProps> & BreadcrumbPublicProps
>
export type BreadcrumbInstance = InstanceType<DefineComponent<BreadcrumbProps>>

export const breadcrumbItemProps = {
  separator: String,
} as const

export type BreadcrumbItemProps = ExtractInnerPropTypes<typeof breadcrumbItemProps>
export type BreadcrumbItemPublicProps = ExtractPublicPropTypes<typeof breadcrumbItemProps>
export type BreadcrumbItemComponent = DefineComponent<
  Omit<HTMLAttributes, keyof BreadcrumbItemPublicProps> & BreadcrumbItemPublicProps
>
export type BreadcrumbItemInstance = InstanceType<DefineComponent<BreadcrumbItemProps>>
