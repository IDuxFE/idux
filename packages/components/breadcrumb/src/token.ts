/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

import type { BreadcrumbProps } from './types'
import type { ComputedRef, InjectionKey, Ref } from 'vue'

export interface BreadcrumbContext {
  mergedPrefixCls: ComputedRef<string>
  separatorRef: Ref<BreadcrumbProps['separator']>
}

export const breadcrumbToken: InjectionKey<BreadcrumbContext> = Symbol('breadcrumbToken')
