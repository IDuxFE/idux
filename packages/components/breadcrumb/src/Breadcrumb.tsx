/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

import { computed, defineComponent, provide, toRef } from 'vue'

import { useGlobalConfig } from '@idux/components/config'

import { breadcrumbToken } from './token'
import { breadcrumbProps } from './types'

export default defineComponent({
  name: 'IxBreadcrumb',
  props: breadcrumbProps,
  setup(props, { slots }) {
    const common = useGlobalConfig('common')
    const mergedPrefixCls = computed(() => `${common.prefixCls}-breadcrumb`)

    provide(breadcrumbToken, {
      mergedPrefixCls,
      separatorRef: toRef(props, 'separator'),
    })

    return () => (
      <nav class={mergedPrefixCls.value} aria-label="Breadcrumb">
        <ol>{slots.default?.()}</ol>
      </nav>
    )
  },
})
