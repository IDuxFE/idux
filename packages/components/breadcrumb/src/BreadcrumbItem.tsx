/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

import { defineComponent, inject } from 'vue'

import { Logger } from '@idux/cdk/utils'

import { breadcrumbToken } from './token'
import { breadcrumbItemProps } from './types'

export default defineComponent({
  name: 'IxBreadcrumbItem',
  props: breadcrumbItemProps,
  setup(props, { slots }) {
    const breadcrumbInjection = inject(breadcrumbToken, null)
    if (!breadcrumbInjection) {
      if (__DEV__) {
        Logger.error('components/breadcrumb', '`breadcrumb-item` must be placed inside `breadcrumb`.')
      }
      return () => null
    }

    const { mergedPrefixCls, separatorRef } = breadcrumbInjection
    return () => (
      <li class={`${mergedPrefixCls.value}-item`}>
        <span class={`${mergedPrefixCls.value}-item-link`}>{slots.default?.()}</span>
        <span class={`${mergedPrefixCls.value}-item-separator`} aria-hidden="true">
          {slots.separator ? slots.separator() : (props.separator ?? separatorRef.value)}
        </span>
      </li>
    )
  },
})
