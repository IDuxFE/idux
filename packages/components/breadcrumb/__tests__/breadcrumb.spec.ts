import { MountingOptions, mount } from '@vue/test-utils'
import { h } from 'vue'

import { renderWork } from '@tests'

import Breadcrumb from '../src/Breadcrumb'
import BreadcrumbItem from '../src/BreadcrumbItem'
import { BreadcrumbProps } from '../src/types'

const defaultSlots = [
  h(BreadcrumbItem, null, { default: () => 'item1' }),
  h(BreadcrumbItem, null, { default: () => 'item2' }),
  h(BreadcrumbItem, null, { default: () => 'item3' }),
]

describe('Breadcrumb', () => {
  const BreadcrumbMount = (options?: MountingOptions<Partial<BreadcrumbProps>>) => {
    const { slots, ...rest } = options || {}
    const mergedOptions = {
      slots: { default: () => defaultSlots, ...slots },
      ...rest,
    } as MountingOptions<BreadcrumbProps>
    return mount(Breadcrumb, mergedOptions)
  }

  renderWork<BreadcrumbProps>(Breadcrumb)

  test('BreadcrumbItem not placed inside Breadcrumb', () => {
    const mockErrorLogger = jest.spyOn(console, 'error').mockImplementation()
    const wrapper = mount(BreadcrumbItem)

    expect(wrapper.isVisible()).toBe(false)
    expect(mockErrorLogger).toBeCalled()
  })

  test("should work with Breadcrumb's `separator` prop", async () => {
    const wrapper = BreadcrumbMount({
      props: {
        separator: '>',
      },
    })
    const separatorDomList = wrapper.findAll('.ix-breadcrumb-item-separator')
    expect(separatorDomList.every(i => i.text() === '>')).toBe(true)
  })

  test("should work with BreadcrumbItem's `separator` prop", async () => {
    const wrapper = BreadcrumbMount({
      slots: {
        default: () => [
          h(BreadcrumbItem, { separator: '>' }, { default: () => 'item1' }),
          h(BreadcrumbItem, null, { default: () => 'item2' }),
        ],
      },
    })
    const separatorDomList = wrapper.findAll('.ix-breadcrumb-item-separator')
    expect(separatorDomList[0].text()).toBe('>')
    expect(separatorDomList[1].text()).toBe('/')
  })

  test("should work with BreadcrumbItem's `separator` slot", async () => {
    const wrapper = BreadcrumbMount({
      slots: {
        default: () => [
          h(BreadcrumbItem, null, { default: () => 'item2' }),
          h(BreadcrumbItem, null, { default: () => 'item2', separator: () => '>' }),
        ],
      },
    })
    const separatorDomList = wrapper.findAll('.ix-breadcrumb-item-separator')
    expect(separatorDomList[0].text()).toBe('/')
    expect(separatorDomList[1].text()).toBe('>')
  })
})
