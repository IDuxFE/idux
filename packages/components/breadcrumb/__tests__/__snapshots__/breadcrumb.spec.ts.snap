// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Breadcrumb > render work 1`] = `
"<nav class="ix-breadcrumb" aria-label="Breadcrumb">
  <ol>
    <!---->
  </ol>
</nav>"
`;
